xdsl/builder.py

```diff
@@ -107,6 +127,113 @@ def region(
         else:
             return Builder._region_args(input)
 
+    @staticmethod
+    def _implicit_region_no_args(func: Callable[[], None]) -> Region:
+        """
+        Generates a single-block region.
+        """
+        block = Block()
+        builder = Builder(block)
+
+        with Builder._push_implicit_builder(builder):
+            func()
+
+        return Region.from_block_list([block])
+
+    @staticmethod
+    def _implicit_region_args(
+        input_types: list[Attribute] | ArrayAttr[Attribute]
+    ) -> Callable[[_CallableImplicitRegionFuncType], Region]:
+        """
+        Decorator for constructing a single-block region, containing the implementation of a
+        function with some input arguments.
+        """
+
+        if isinstance(input_types, ArrayAttr):
+            input_type_seq = input_types.data
+        else:
+            input_type_seq = input_types
+
+        def wrapper(func: _CallableImplicitRegionFuncType) -> Region:
+            block = Block.from_arg_types(input_type_seq)
+            builder = Builder(block)
+
+            with Builder._push_implicit_builder(builder):
+                func(block.args)
+
+            region = Region.from_block_list([block])
+            return region
+
+        return wrapper
+
+    @overload
+    @staticmethod
+    def implicit_region(
+        input: list[Attribute] | ArrayAttr[Attribute]
+    ) -> Callable[[_CallableImplicitRegionFuncType], Region]:
+        """
+        Annotation used to construct a Region tuple from a function.
+        The annotation can be used in two ways:
+
+        For regions that have inputs or outputs:
+        ```
+        @Builder.implicit_region(input_types)
+        def func(args: tuple[BlockArgument, ...]) -> None:
```

Comment:
Same question as last time I think, could we inline the `tuple`?
So have `def func(arg1, arg2, ...)` instead of a tuple?

Reply:
There's no straightforward way of doing it directly, since the tuple splat is not a normal argument there's no way to specify it in the Callable type signature. There might be a way to hack ParamSpec to make it work but I think the issue was that it would silently ignore it if you gave the wrong types.